goenv/version.go

```diff
@@ -4,29 +4,47 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"runtime/debug"
 	"strings"
 )
 
 // Version of TinyGo.
 // Update this value before release of new version of software.
 const version = "0.35.0-dev"
 
-var (
-	// This variable is set at build time using -ldflags parameters.
-	// See: https://stackoverflow.com/a/11355611
-	GitSha1 string
-)
-
 // Return TinyGo version, either in the form 0.30.0 or as a development version
 // (like 0.30.0-dev-abcd012).
 func Version() string {
 	v := version
-	if strings.HasSuffix(version, "-dev") && GitSha1 != "" {
-		v += "-" + GitSha1
+	if strings.HasSuffix(version, "-dev") {
+		if hash := readGitHash(); hash != "" {
+			v += "-" + hash
+		}
 	}
 	return v
 }
 
+func readGitHash() string {
+	hash := ""
+	modified := false
+	if info, ok := debug.ReadBuildInfo(); ok {
+		for _, setting := range info.Settings {
+			println("setting:", setting.Key, setting.Value)
+			switch setting.Key {
+			case "vcs.revision":
+				hash = setting.Value[:8]
+			case "vcs.modified":
+				modified = setting.Value == "true"
```

Comment:
If possible, I'd like the state to be unmodified when building from CI. Note that `vcs.modified` will still return true if there are any untracked files present.

```
>.\bin\tinygo.exe version
.....
setting: vcs git
setting: vcs.revision e5f95be5d00ceb83ddce36374bfb06ba81c9264e
setting: vcs.time 2024-11-03T09:08:14Z
setting: vcs.modified true
tinygo version 0.35.0-dev-e5f95be5-modified windows/amd64 (using go .\bin\tinygo.exe versionM version 18.1.2)

```